action.php

```diff
@@ -1,49 +1,111 @@
+     * @return void
+     */
+    public function register(Doku_Event_Handler $controller)
+    {
         global $conf;
-        if ($conf['authtype'] == 'authdiscourse') {
-            $controller->register_hook('HTML_LOGINFORM_OUTPUT', 'BEFORE', $this, 'handle_login_form');
+        if ($conf['authtype'] !== 'authdiscourse') return;
+
+        $controller->register_hook('HTML_LOGINFORM_OUTPUT', 'BEFORE', $this, 'handle_loginform'); // old
+        $controller->register_hook('FORM_LOGIN_OUTPUT', 'BEFORE', $this, 'handle_loginform'); // new
+
+        $controller->register_hook('ACTION_ACT_PREPROCESS', 'BEFORE', $this, 'handle_dologin');
+        $controller->register_hook('ACTION_ACT_PREPROCESS', 'BEFORE', $this, 'handle_dologout');
+    }
+
+    /**
+     * Replaces login form with a button
+     *
+     * @param Doku_Event $event the form event
+     * @param mixed $param unused
+     * @return void
+     */
+    public function handle_loginform(Doku_Event $event, $param)
+    {
+        global $ID;
+        $html = '<a href="' . wl($ID, ['do' => 'login']) . '" title="Discourse">';
+        $html .= inlineSVG(__DIR__ . '/images/discourse.svg', 6000);
+        $html .= '</a>';
+        $form = $event->data;
+
+        if (is_a($form, \dokuwiki\Form\Form::class)) {
+            // New: clear the form
+            for ($i = $form->elementCount(); $i > 0; $i--) {
+                $form->removeElement($i - 1);
+            }
+            $form->addFieldsetOpen($this->getLang('login_with'))->addClass('plugin_authdiscourse');
+            $form->addHTML($html);
+            $form->addFieldsetClose();
+        } else {
+            // Old: replace form content
+            $form->_content = [];
+            $form->_content[] = form_openfieldset([
+                '_legend' => $this->getLang('login_with'),
+                'class' => 'plugin_authdiscourse',
+            ]);
+            $form->_content[] = $html;
+            $form->_content[] = form_closefieldset();
         }
     }
 
-	function handle_login_form(&$event, $param)
-	{
-        global $auth, $lang;
+    /**
+     * Redirects to auth url
+     *
+     * @param Doku_Event $event
+     * @param $param
+     */
+    public function handle_dologin(Doku_Event $event, $param)
+    {
+        if ($event->data !== 'login') {
+            return;
+        }
 
-        $loginurl = $auth->getLoginURL();
+        global $auth;
 
-        // Replace the whole existing form as we can't handle username/password.
-        $event->data->_content = array();
-        $event->data->insertElement(0, '<a href="'.$loginurl.'" style="border: 5px solid; font-size: 200%; padding: 0.2em 0.5em;">'.$lang['btn_login'].'</a>');
-	}
+        $endpoint = rtrim($this->getConf('sso_url'), '/');
+        if (!preg_match('!/session/sso_provider$!', $endpoint)) $endpoint .= '/session/sso_provider';
+        $secret = $this->getConf('sso_secret');
+        $nonce = md5(random_bytes(18));
```

Comment:
Why the call to `md5` here? I think we should just encode the nonce as base64, as it was before.

Reply:
MD5 was used here as a simple method to ensure the nonce is ASCII only. But you're right, we're loosing two bytes of randomness here. I changed it to base64. (even though 16 vs 18 bytes probably do not matter much here)